cpp/src/cores/trim.cu

```diff
@@ -0,0 +1,119 @@
+#include <cugraph/utilities/error.hpp>
+#include <cugraph/utilities/host_scalar_comm.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/error.hpp>
+
+
+namespace cugraph {
+template <typename vertex_t>
+struct extract_zero_core_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(vertex_t src,
+                                                                            vertex_t dst,
+                                                                            uint8_t src_in_zero_core,
+                                                                            uint8_t dst_in_zero_core,
+                                                                            thrust::nullopt_t) const
+  {
+    return (src_in_zero_core == uint8_t{0}) && (dst_in_zero_core == uint8_t{0})
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : thrust::nullopt;
+  }
+};
+
+template <typename edge_t>
+struct is_zero_or_greater_t {
+  __device__ uint8_t operator()(edge_t core_number) const
+  {
+    return core_number >= edge_t{0} ? uint8_t{1} : uint8_t{0};
+  }
+};
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+rmm::device_uvector<vertex_t> 
+trim(raft::handle_t const& handle,
+       graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view)
+{
+//    std::optional<rmm::device_uvector<vertex_t>> renumber_map{std::nullopt};
+//    auto cur_graph_view = modified_graph_view ? *modified_graph_view : graph_view;
+//    auto vertex_partition_range_lasts =
+//      renumber_map?
+//         std::make_optional<std::vector<vertex_t>>(graph_view.vertex_partition_range_lasts())
+//        : std::nullopt;
+    std::optional<graph_t<vertex_t, edge_t, false, multi_gpu>> modified_graph{std::nullopt};
+    std::optional<graph_view_t<vertex_t, edge_t, false, multi_gpu>> modified_graph_view{std::nullopt};
+    std::optional<rmm::device_uvector<vertex_t>> renumber_map{std::nullopt};
+    auto vertex_partition_range_lasts =
+         std::make_optional<std::vector<vertex_t>>(graph_view.vertex_partition_range_lasts());
+
+    rmm::device_uvector<edge_t> core_numbers(graph_view.number_of_vertices(), handle.get_stream());
+
+    core_number(
+      handle, graph_view, core_numbers.data(), k_core_degree_type_t::OUT, size_t{0}, size_t{0});
+
+    edge_src_property_t<decltype(graph_view), uint8_t> edge_src_in_zero_cores(handle,
+                                                                                 graph_view);
+    edge_dst_property_t<decltype(graph_view), uint8_t> edge_dst_in_zero_cores(handle,
+                                                                                 graph_view);
+    auto in_zero_core_first =
+      thrust::make_transform_iterator(core_numbers.begin(), is_zero_or_greater_t<edge_t>{});
+    rmm::device_uvector<uint8_t> in_zero_core_flags(core_numbers.size(), handle.get_stream());
```

Comment:
Why are you finding 0-core? Vertex degrees can't be negative, so 0-core is the entire graph.

Here, we are trying to extract a subgraph where every vertex has non-zero in&out degree (as zero indegree or out degree vertices are singleton components).

You can do something like the following.

```
auto in_degrees = graph_view.compute_in_degrees(handle);
auto out_degrees = graph_view.compute_out_degrees(handle);
rmm::device_uvector<uint8_t> include_flags(in_degrees.size(), handle.get_stream());
auto degree_pair_first = thrust::make_zip_iterator(thrust::make_tuple(in_degrees.begin(), out_degrees.begin()));
thrust::transform(handle.get_thrust_policy(), degree_pair_first,  degree_pair_last, []__device__(auto pair) {
  return (thrust::get<0>(pair) >= 1 && thrust::get<1>(pair) >= 1) ? uint8_t{1} : uint8_t{0};
});
edge_src_property_t<decltype(graph_view), uint8_t> edge_src_include_flags(handle, graph_view);
edge_dst_property_t<decltype(graph_view), uint8_t> edge_dst_include_flags(handle, graph_view);
update_edge_src_property(...);
update_edge_dst_property(...);
```